orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1119,10 +1137,35 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceSumGradient) {
   }
 
   ArgDef grad = GO(0);
-  if (!keepdims && attributes.find("axes") != attributes.end()) {
-    std::vector<int64_t> axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
-    grad = IA("Unqueezed_Grad");
-    result.push_back(NodeDef("Unsqueeze", {GO(0)}, {grad}, {MakeAttribute("axes", axes_values)}));
+  if (!keepdims) {
+    size_t numInputs = GetSrcNodeInputSize();
+    if (numInputs == 1) {  //axes is attribute
+      if (attributes.find("axes") != attributes.end()) {
+        std::vector<int64_t> axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+
+        grad = IA("Unqueezed_Grad");
+        result.push_back(NodeDef("Unsqueeze", {GO(0)}, {grad}, {MakeAttribute("axes", axes_values)}));
+
+      } else {  //axes is not available, try to add grad graph from input shape info
+        std::vector<Dimension> input_shape;
+        if (GetShape(I(0), input_shape).IsOK()) {
+          size_t nDim = input_shape.size();
+          std::vector<int64_t> axes_values(nDim);
+          std::iota(axes_values.begin(), axes_values.end(), 0);
+          grad = IA("Unqueezed_Grad");
+          result.push_back(NodeDef("Unsqueeze", {GO(0)}, {grad}, {MakeAttribute("axes", axes_values)}));
+        } else {
+          ORT_THROW(
+              "Cannot add gradient graph as attribute 'axes' "
+              "is not set and input shape is not available.");
+        }
+      }
```

Comment:
not sure if this piece of code is needed. 
if axes is empty, ReduceSum will reduce to single scalar. 
I think latter Expand op can always operate on scalar without problem. so the Unsqueezeis actually not needed